processor/src/advice/mod.rs

```diff
@@ -1,12 +1,63 @@
 use super::{ExecutionError, Felt, ProgramInputs, Word};
+use core::mem;
 use vm_core::{
+    errors::InputError,
     utils::{
         collections::{BTreeMap, Vec},
         IntoBytes,
     },
     AdviceSet, StarkField,
 };
 
+pub trait AdviceProvider {
+    // ADVICE TAPE
+    // --------------------------------------------------------------------------------------------
+
+    // TODO this looks like a `Self: Iterator<Item = Felt>`
+    fn read_tape(&mut self) -> Result<Felt, ExecutionError>;
+    fn read_tape_w(&mut self) -> Result<Word, ExecutionError>;
+    fn read_tape_dw(&mut self) -> Result<[Word; 2], ExecutionError>;
+
+    // TODO this looks like a `Self: Extend<Felt>`
+    fn write_tape(&mut self, value: Felt);
+    fn write_tape_from_map(&mut self, key: Word) -> Result<(), ExecutionError>;
+
+    // TODO consider creating an `extend_into_map` so the user can extend a given set for a given
+    // key
+    fn insert_into_map(&mut self, key: Word, values: Vec<Felt>) -> Result<(), ExecutionError>;
+
+    // ADVISE SETS
+    // --------------------------------------------------------------------------------------------
+
+    fn get_tree_node(&self, root: Word, depth: Felt, index: Felt) -> Result<Word, ExecutionError>;
+    fn get_merkle_path(
+        &self,
+        root: Word,
+        depth: Felt,
+        index: Felt,
+    ) -> Result<Vec<Word>, ExecutionError>;
+
+    // TODO the compact merkle tree will take depth as argument so leaves can be inserted on
+    // arbitrary depth. this design doesn't support that, so we likely need to add a `depth`
+    // parameter in the future.
+    fn update_merkle_leaf(
+        &mut self,
+        root: Word,
+        index: Felt,
+        leaf_value: Word,
+        update_in_copy: bool,
+    ) -> Result<Vec<Word>, ExecutionError>;
+
+    // CONTEXT MANAGEMENT
+    // --------------------------------------------------------------------------------------------
+
+    // TODO review if we are not enforcing some external control that should be internal to the VM.
+    // example: if the read_tape fails, the provider sends the VM the current step so it can keep
+    // track of where the error happened. however, this seems to be a responsibility of the VM as
+    // it should also know what is the step that caused the failure.
+    fn advance_clock(&mut self);
```

Comment:
I believe we need to keep track of the clock cycle primarily for more informative errors. I agree, there should be a better solution to this and we should get rid of this function - but I would leave it to future PRs.

Reply:
Sounds good! I'm thinking of leaving these TODO notes; wdyt?

Reply:
Yep - that works!